lib/history/sqlite/sqlite.go

```diff
@@ -196,13 +165,36 @@ func (t *Timeline) RecordTimeline(ctx context.Context, events []*pb.TimelineEven
 		filtered = append(filtered, event)
 	}
 
-	if err = t.insertEvents(ctx, filtered); err != nil {
+	if err := t.insertEvents(ctx, filtered); err != nil {
 		return trace.Wrap(err)
 	}
 
 	return nil
 }
 
+// insertEvents inserts the provided events into the timeline.
+func (t *Timeline) insertEvents(ctx context.Context, events []*pb.TimelineEvent) error {
+	sqlExecer := newSQLExecer(t.database)
+	for _, event := range events {
+		row, err := newDataInserter(event)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		err = row.Insert(ctx, sqlExecer)
+		// Unique constraint error indicates duplicate row.
+		// Just ignore duplicates and continue.
+		if isErrConstraintUnique(err) {
+			continue
+		}
+		if err != nil {
+			return trace.Wrap(err)
```

Comment:
Should this method attempt to insert all events rather than bailing on first error? Or are they gonna be retried later?

Reply:
```golang
// UpdateTimeline updates the timeline with a new event.
// Duplicate requests will have no effect.
func (r *server) UpdateTimeline(ctx context.Context, req *pb.UpdateRequest) (*pb.UpdateResponse, error) {
	if err := r.agent.RecordTimeline(ctx, []*pb.TimelineEvent{req.GetEvent()}); err != nil {
		return nil, GRPCError(err)
	}
	if err := r.agent.RecordLastSeen(req.GetName(), req.GetEvent().GetTimestamp().ToTime()); err != nil {
		return nil, GRPCError(err)
	}
	return &pb.UpdateResponse{}, nil
}
```
If the agent fails to record all events, the last seen timestamp will not be updated and they will be retried next time.
Maybe `RecordTimeline` should at least return the timestamp of the last event it recorded.